include/openssl/kdf.h

```diff
@@ -83,6 +83,29 @@ OPENSSL_EXPORT int EVP_PKEY_CTX_add1_hkdf_info(EVP_PKEY_CTX *ctx,
                                                const uint8_t *info,
                                                size_t info_len);
 
+// SSH-specific KDF
+//
+// This KDF should only be called from SSH client/server code; it's not a
+// general-purpose KDF and is only Approved for FIPS 140-3 use specifically
+// in SSH.
+
+// |type| values for SSHKDF().
+#define EVP_KDF_SSHKDF_TYPE_INITIAL_IV_CLI_TO_SRV     65
+#define EVP_KDF_SSHKDF_TYPE_INITIAL_IV_SRV_TO_CLI     66
+#define EVP_KDF_SSHKDF_TYPE_ENCRYPTION_KEY_CLI_TO_SRV 67
+#define EVP_KDF_SSHKDF_TYPE_ENCRYPTION_KEY_SRV_TO_CLI 68
+#define EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_CLI_TO_SRV  69
+#define EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_SRV_TO_CLI  70
+
+// SSHKDF calculates |out_len| bytes of the SSH KDF, using |digest|, and
+// writes them to |out|. It returns one on success and zero on error.
+OPENSSL_EXPORT int SSHKDF(const EVP_MD *evp_md,
+                          const uint8_t *key, size_t key_len,
+                          const uint8_t *xcghash, size_t xcghash_len,
+                          const uint8_t *session_id, size_t session_id_len,
```

Comment:
Maybe it's worth introducing a new struct that has `uint8_t*` and `size_t` and then use that to reduce the number of arguments. That would increase readability of the API as well.

Reply:
We already have that, it's a crypto byte string [cbs_st](https://github.com/awslabs/aws-lc/blob/main/include/openssl/bytestring.h#L40), it is primarily used for internal functions/parsing in libssl. The convention for libcrypto's public API is unfortunately lots of pointers and lengths which offers the greatest flexibility for other languages and consumers at the cost of complexity and easy to misuse API.

What API does OpenSSL offer for this KDF? Is that the API OpenSSH expects?

Reply:
This is the pattern used by the rest of the APIs... we generally do what BoringSSL/OpenSSL do, so we're more compatible with existing Boring/Open SSL code, and users of things we've added have a familiar API.

Reply:
OpenSSL has an EVP API for this; the `SSHKDF()` function there is `static`. This seemed like the least intrusive way of adding this functionality.

Reply:
@andrewhop Similar to symmetric encryption EVP APIs, OpenSSL offers EVP_KDF_ APIs and as a matter of fact, OpenSSL supports SSHKDF: https://www.openssl.org/docs/man3.0/man7/EVP_KDF-SSHKDF.html. I think it's best to introduce this API via the existing framework.

Reply:
The EVP_KDF API is so terrible to use though. I suppose we'll have to though, if our goal is to remain compatible with OpenSSL.

Reply:
The public EVP_KDF API and internals is not compatible with AWS-LC for many reasons and we are not interested in pulling in all that code. Especially not for something as esoteric as an SSH KDF.

I think current approach is good.

Reply:
I second Torben. See my response to the above comment: https://github.com/awslabs/aws-lc/pull/656#discussion_r1001040100